### PR TITLE
Support "." char as user name for User/Orgs in RSS/ATOM/GPG/KEYS path ...

### LINKED ISSUE #22301: [Bug] RSS URI Access Encoding Issues
### Description

Attempting to access the RSS feed of a user that has a period in their name results in a 404. Encoding the URI with %2e works. The problem exists on try.gitea.io and my own self-hosted Gitea, as well as other publicly-facing Gitea instances.

### Gitea Version

1.17.2

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

N/A

### Screenshots

_No response_

### Git Version

N/A

### Operating System

Linux, try.gitea.io

### How are you running Gitea?

Various means, sometimes stand-alone, sometimes other sites.

### Database

PostgreSQL

### PR DESCRIPTION
**Pull request by 6543**
- close #22301

workaround for https://github.com/go-chi/chi/issues/781

### CODE REVIEW

File: routers/web/web.go [hunk 1]
```diff
@@ -672,16 +673,43 @@ func RegisterRoutes(m *web.Route) {
...
-			m.Get(".keys", user.ShowSSHKeys)
-			m.Get(".gpg", user.ShowGPGKeys)
-			m.Get(".rss", feedEnabled, feed.ShowUserFeedRSS)
-			m.Get(".atom", feedEnabled, feed.ShowUserFeedAtom)
-			m.Get("", user.Profile)
-		}, func(ctx *context.Context) {
-			ctx.Data["EnableFeed"] = setting.EnableFeed
-		}, context_service.UserAssignmentWeb())
+		m.Get("/{username}", func(ctx *context.Context) {
+			// WORKAROUND to support usernames with "." in it
```
Comment (6543) [inline routers/web/web.go:678]: if somebody do have a better workaround please tell :/

File: build.go [hunk 2]
```diff
@@ -1,7 +1,6 @@
 // Copyright 2020 The Gitea Authors. All rights reserved.
 // SPDX-License-Identifier: MIT
 
-
```
Comment (yardenshoham) [inline build.go]: Unrelated

Comment (6543) [reply build.go]: It is ... but not worth it's own pull

**Comment by 6543** [comment]
@yardenshoham why did you remove the backport if  #22301 was assigned to v1.18.6 ?!?

File: routers/web/web.go (hunk 1 above)
Comment (wxiaoguang) [reply routers/web/web.go:678]: I guess this design would cause problem in the future.

For example, what if a user/org is called `server.ssh` or `mylibrary.png` ?

Should we introduce a breaking change to move these suffixes to a never-conflicting sub path?

Comment (6543) [reply routers/web/web.go:678]: @wxiaoguang that's how it actually works atm!

this just fix an edgecase where the name is not "abc" but e.g. "ab.c" for example

Comment (6543) [reply routers/web/web.go:678]: https://github.com/go-gitea/gitea/blob/03591f0f95823a0b1dcca969d2a3ed505c7e6d73/models/user/user.go#L540

Comment (wxiaoguang) [reply routers/web/web.go:678]: But I see a ".png" there .... which is not in the reservedUserPatterns

Comment (6543) [reply routers/web/web.go:678]: I also agree that we should not add more to that list ...

Comment (6543) [reply routers/web/web.go:678]: @wxiaoguang well we just discovered an edgecase ... we could either drop one and break things or add it and might break another thing ...

Comment (wxiaoguang) [reply routers/web/web.go:678]: I would suggest to drop the `.png` here. Users could use `https://try.gitea.io/user/avatar/wxiaoguang/0` to fetch avatar image (well, this URL could also be fine-tuned). The `.png` seems quite misleading too, why not `.jpg` ? 😁 .

And maybe it's better to leave a big warning comment here to prevent future developers from adding more extensions.

**Comment by zeripath** [comment]
wouldn't `m.Get("/{username}.png", func(ctx *context.Context) {` etc. work?

File: models/user/user.go [hunk 3]
```diff
@@ -537,6 +537,7 @@ var (
 		"gitea-actions",
 	}
 
+	// DON'T ADD ANY NEW STUFF, WE SOLVE THIS WITH ALTERNATE PATHS!
```
Comment (6543) [inline models/user/user.go]:
```suggestion
	// DON'T ADD ANY NEW STUFF, WE SOLVE THIS WITH `/user/{obj}`  PATHS!
```
?

**Comment by 6543** [comment]
> wouldn't `m.Get("/{username}.png", func(ctx *context.Context) {` etc. work?

well that's also affected by the chi issue :/ so NO sadly

**Comment by 6543** [comment]
@yardenshoham  I'm doing a partial backport ... other option would be to split out this pull

File: routers/web/web.go [hunk 4]
```diff
@@ -672,16 +674,39 @@ func RegisterRoutes(m *web.Route) {
...
+				}
+			case strings.HasSuffix(username, ".atom"):
+				feedEnabled(ctx)
+				if !ctx.Written() {
+					ctx.SetParams("username", strings.TrimSuffix(username, ".atom"))
+					context_service.UserAssignmentWeb()(ctx)
+					feed.ShowUserFeedAtom(ctx)
+				}
+			default:
+				context_service.UserAssignmentWeb()(ctx)
```
Comment (6543) [inline routers/web/web.go:711]: uh I need to check for is written ... now it's time to move it in a helper func :/

**Comment by yardenshoham** [comment]
I just want to make sure there won't be any breaking changes backported

**Comment by silverwind** [comment]
> [drop "/{username}.png" in fafour of "/user/avatar/{username}"](https://github.com/go-gitea/gitea/pull/23874/commits/0129cb1ba41c1073cdabd7bc9864f4a878427361)

Do we know any users of this? I checked drone, it uses `/avatars/<hash>`, so is safe.

**Comment by 6543** [comment]
> > [drop "/{username}.png" in fafour of "/user/avatar/{username}"](https://github.com/go-gitea/gitea/pull/23874/commits/0129cb1ba41c1073cdabd7bc9864f4a878427361)
> 
> Do we know any users of this? I checked drone, it uses `/avatars/<hash>`, so is safe.

well in any case it is breaking but we have an alternative route to switch to ...

and we never did any hard constrains on the normal web urls ... only api paths should be backwards compatible

**Comment by 6543** [comment]
partial backport -> #23878

File: models/user/user.go [hunk 5]
```diff
@@ -537,6 +537,7 @@ var (
 		"gitea-actions",
 	}
 
+	// DON'T ADD ANY NEW STUFF, WE SOLVE THIS WITH `/user/{obj}`  PATHS!
 	reservedUserPatterns = []string{"*.keys", "*.gpg", "*.rss", "*.atom"}
```
Comment (silverwind) [inline models/user/user.go]: Maybe remove these then if it can be fully solved via router?

File: models/user/user.go [hunk 6]
```diff
@@ -537,6 +537,7 @@ var (
 		"gitea-actions",
 	}
 
+	// DON'T ADD ANY NEW STUFF, WE SOLVE THIS WITH `/user/{obj}`  PATHS!
```
Comment (silverwind) [inline models/user/user.go]:
```suggestion
	// DON'T ADD ANY NEW STUFF, WE SOLVE THIS WITH `/user/{obj}` PATHS!
```

File: models/user/user.go (hunk 5 above)
Comment (6543) [reply models/user/user.go]: droping already reserved stuff would be another pull ... and we already have reserved it so it's fine just dont add new stuff

Comment (6543) [reply models/user/user.go]: PS: in this regards, we do mimic github witch does the same!

e.g.: https://github.com/silverwind.keys

Comment (silverwind) [reply models/user/user.go]: If `alicia.keys.keys` works, and the router bug is the only reason we reserve, let's drop the reservations.

**Comment by 6543** [comment]
test only run: https://drone.gitea.io/go-gitea/gitea/71240/3/7 to see that the test actually cover this bug :)

**Comment by wxiaoguang** [comment]
I think it LGTM (although my LGTM doesn't count).

Some small questions:

* What's the test fixture `action.id=9` used for?
* Could an existing user be re-used for the testing, then no need to introduce a new `user34`? It could help to avoid code bloating.

File: routers/web/feed/profile.go [hunk 7]
```diff
@@ -39,10 +41,22 @@ func showUserFeed(ctx *context.Context, formatType string) {
 		return
 	}
 
+	ctxUserDescription, err := markdown.RenderString(&markup.RenderContext{
```
Comment (lunny) [inline routers/web/feed/profile.go]: This is an unrelated change?

Comment (wxiaoguang) [reply routers/web/feed/profile.go]: The description: `render description for User/Org rss`

On web UI, it's also rendered as markdown IIRC.

Comment (6543) [reply routers/web/feed/profile.go]: It's also a related rss bugfix ... but I can split it out (do not have an issue id jet)

**Comment by 6543** [comment]
> Could an existing user be re-used for the testing, then no need to introduce a new user34? It could help to avoid code bloating.

I would have to touch fare more code if i do use an existing user in the fixtures and change it's name, and so I did add an user witch cover all cases about naming and explicit add an action entry (indepentent from existing activity tests) and made it valid, so rss can parse and convert it.

File: routers/web/feed/profile.go (hunk 7 above)
Comment (6543) [reply routers/web/feed/profile.go:44]: Y/N ?

Comment (6543) [reply routers/web/feed/profile.go]: -> #23887

**Comment by 6543** [comment]
blocked by #23887

**Comment by wxiaoguang** [comment]
> blocked by #23887

No block. #23887 only fixes an old bug, not related to this one 😁

**Comment by 6543** [comment]
well but it conflicts with the test ... so I just wait for it to be merged by merge main into here, the fix is gone and test passes

so well it blocks ... or I remove it here now and fix conflicts later ...

File: routers/web/web.go [hunk 8]
```diff
@@ -672,16 +674,43 @@ func RegisterRoutes(m *web.Route) {
...
-			m.Get(".png", user.AvatarByUserName)
-			m.Get(".keys", user.ShowSSHKeys)
-			m.Get(".gpg", user.ShowGPGKeys)
-			m.Get(".rss", feedEnabled, feed.ShowUserFeedRSS)
-			m.Get(".atom", feedEnabled, feed.ShowUserFeedAtom)
-			m.Get("", user.Profile)
-		}, func(ctx *context.Context) {
-			ctx.Data["EnableFeed"] = setting.EnableFeed
-		}, context_service.UserAssignmentWeb())
+		m.Get("/{username}", func(ctx *context.Context) {
```
Comment (zeripath) [inline routers/web/web.go:677]: OK, I think this function should get moved out of web.go and into a named function within routers/web/user - and we should probably also update the `routing.UpdateFuncInfo(req.Context(), ...)` (Although we should do that without causing the funcInfo to be generated at run time.)

Comment (6543) [reply routers/web/web.go:677]: In the best case the upstream issue get solved ...

the question is if we wana have that workaround now in or engineer around it properly ... or wait for upstream

**Comment by 6543** [comment]
moved breaking part into  #23908

File: routers/web/web.go (hunk 8 above)
Comment (silverwind) [reply routers/web/web.go:677]: I have personally given up on waiting for `chi` to fix any issues 😉

**Review by silverwind — APPROVED** [review]

File: routers/web/web.go (hunk 8 above)
Comment (6543) [reply routers/web/web.go:677]: well that would mean **we** must fix and submit a patch upstream :laughing:

**Review by lunny — APPROVED** [review]
